### PR TITLE
WIP: Redo sherpa installation and CI with PEP 518 and tox

### PR DESCRIPTION
**Pull request by hamogu**
In this PR, I'm changing sherpa's setup scripts and CI setup to follow PEP 518 and astropy's examples.
**This is not functional right now.** I'm posting it here to get some feedback from the Sherpa developers to see if this is a route they are interested in. Getting it all right, especially with the binary components requires some work and probably help from core developers.

This largely follows what astropy does.

- [x] replace versioneer with setuptools_scm (not strictly required, but since this relies on setuptools now seems like a small step to take)
- [x] Install python files
- [ ] compile and install extensions
- [x] run tests for python with tox 
- [ ] run tests using compiled extensions with tox
- [ ] run CI with tox
- [ ] write / update contributor docs (currently referring to astropy docs)

### CODE REVIEW

**Comment by DougBurke** [comment]
I'm going to close #736 as the idea there was to work out how to update the build infrastructure, but I don't know how.

**Comment by DougBurke** [comment]
I personally am interested in this work, but progress will be slow here as our build team is swamped with other stuff.

One thing I don't have a good handle on is what the current best practices are for python development and testing nowadays (e.g. tox is mentioned as a replacement for 'python setup.py test`, but what other options are there?). There's also a number of build-related issues that are presumably related to the fact that our build system was put together a long time ago; for instance #50 #198 #480 #595 (a rather incomplete list)

**Comment by hamogu** [comment]
@DougBurke Ifyou and I are interested, that might be enough to get this developed. I just have no experience with the compiled modules in sherpa. I reviewed the issues you linked and it seems that they could all be solved by this. At the same time, I noticed that the CI matrix is a little "odd", e.g. we are testing "python setup.py install" and "python setup.py develop" separately, but on the other hand, we have very little coverage in terms of python / numpy / astropy versions.

All I need to continue at this point is a "Yes, there is a realistic chance this will get merged eventually". I just don't want to spend days on something that has no chance of being accepted.


My draft is taken from applying what astropy does - I think there is no shame to take from a project with a much larger pool of developers that explicitly puts time, effort (and money) into developing build tools and documentation for how to build with the expressed goal to make live easier for other scientific python packages.

A few links for my own notes:

- https://packaging-guide.openastronomy.org/en/latest/docs.html
- https://github.com/pypa/manylinux
- https://github.com/joerick/cibuildwheel/
- https://openastronomy-azure-pipelines.readthedocs.io/en/latest/ (which uses the two packages above)

**Comment by hamogu** [comment]
A few notes for me to look at:

- https://packaging-guide.openastronomy.org/en/latest/
- An example using cibuildwheel behind the scenes: https://openastronomy-azure-pipelines.readthedocs.io/en/latest/publish.html 
- https://github.com/pypa/manylinux/
- https://github.com/joerick/cibuildwheel/

That does not mean that I intend to use those tools, but that I should look and see if we should do that or get some ideas to maybe not at all.

File: setup.py [hunk 1]
```diff
@@ -18,131 +18,67 @@
...
-        "Other dependencies will be automatically installed\n"
-        "Please install NUMPY (e.g. pip install numpy) and try again."
-    ), file=sys.stderr)
-    sys.exit(2)
+if 'test' in sys.argv:
+    print(TEST_HELP)
+    sys.exit(1)
 
-from helpers.extensions import static_ext_modules
+DOCS_HELP = """
```
Comment (DougBurke) [inline setup.py:52]: A minor issue, but the existing documentation only talks about using `build_sphinx` and not `build_docs` - i.e. https://sherpa.readthedocs.io/en/latest/install.html#building-the-documentation

Do we want this message to note the `build_sphinx` option as well?

File: setup.cfg [hunk 2]
```diff
@@ -1,10 +1,68 @@
...
-install = sherpa_config xspec_config install
-test = develop test
-bdist_wheel = sherpa_config bdist_wheel
+[metadata]
+name = sherpa
+author = Smithsonian Astrophysical Observatory / Chandra X-Ray Center
+author_email = cxchelp@head.cfa.harvard.edu
+license = GNU GPL v3
+license_file = COPYRIGHT
+url = http://cxc.harvard.edu/sherpa/
```
Comment (DougBurke) [inline setup.cfg:7]: We should probably change this to the rtd page? If nothing else we should change to the https version ;-)

File: setup.cfg [hunk 3]
```diff
@@ -103,3 +161,42 @@ source-dir = docs
 #wcslib_libraries = wcs
 #gfortran_lib_dirs = None
 #gfortran_libraries = gfortran
+
+
+[coverage:run]
+omit =
```
Comment (DougBurke) [inline setup.cfg:167]: I'm guessing we need to replace `astropy` by `sherpa`, and probably tweak the directories?

Comment (hamogu) [reply setup.cfg:167]: Yes. I don't know if we've set up any coverage testing for Sherpa at all, so this will need some thought.

**Comment by DougBurke** [comment]
Given that #782 has been merged can this be rebased now?

**Comment by hamogu** [comment]
Sure. But there are still a number of steps missing, this is very much a WIP. It will probably have to be rebased more than once before it's ready, so I don't think there is a hurry.

**Comment by DougBurke** [comment]
We have code that has a mixture of absolute and relative imports - e.g. 

```
from sherpa.bob import bobby
from .fred import freddy
```

I'm wondering if that could be the cause of the mysterious `cannot import from sherpa._utils` message we can see? Clutching at straws, but it was a problem in some of my recent experiments to support the aim if this PR.

**Comment by DougBurke** [comment]
Just to note that in https://github.com/cxcsds/xspec-models-cxc I was able to combine new-style declarative setup.cfg with old-style "build an extension". That's not to say we won't have problems [e.g. the location of the grpup/stk libraries #50 ] but there is hope. And something we should definitely work on sooner rather than later after 4.14 has stabilised

**Pull request closed without merging**